SubMesh support for nonconformal AMR

hughcars opened the pull request:
This PR introduces support for nonconformal amr using SubMesh and ParSubMesh for exterior surface submesh and volume submesh. This PR does not currently support interior surface submeshes, as there is an issue for TransferMap where the resulting dof map may not be an isomorphism (1 to 1). This is almost certainly due to an issue where a master/slave face have different orientation on an internal boundary. 

Additional fixes in this PR:
- Array now has an initializer list constructor
- Boundary attributes for parallel volume submeshes with internal boundaries are now transferred via a "ghost boundary attributes" technique as opposed to an RT0 solution transfer.
- Array supports "ShrinkToFit" which copies an array to a buffer exactly matching the internal size.
- Array has Permute method, which can use an index array to reorder an array in place
- IntegerSet is now derived from Array, many methods are marked const
<!--GHEX{"author":"hughcars","assignment":"2024-09-10T18:45:55","editor":"tzanio","id":4476,"reviewers":["dylan-copeland","jakubcerveny","rw-anderson","jandrej"],"merge":"2024-11-15T21:00:04","approval":"2024-11-09T04:14:03"}XEHG-->
<!--GHEXTABLE-->
 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#4476](https://github.com/mfem/mfem/pull/4476) | @hughcars | @tzanio | @dylan-copeland + @jakubcerveny + @rw-anderson + @jandrej | 9/10/24 | 11/8/24 | 11/15/24 | |
<!--ELBATXEHG-->
<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

tzanio commented:
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

dylan-copeland commented on mesh/pmesh.hpp:
```diff
@@ -446,11 +446,41 @@ class ParMesh : public Mesh
 
+   /**
+    * @brief Accessors for entities within a shared group structure.
+    * @details For all vertex/edge/face the two argument version returns the local index,
```
Long lines. Also, this sentence should be split into two.

dylan-copeland commented on mesh/pmesh.hpp:
```diff
@@ -496,10 +526,6 @@ class ParMesh : public Mesh
 
-   /** Return true if the face is interior or shared. In parallel, this
-       method only works if the face neighbor data is exchanged. */
-   inline bool FaceIsTrueInterior(int FaceNo) const { return Mesh::FaceIsTrueInterior(FaceNo); }
```
Can this be kept for backwards compatibility?

hughcars replied:
The `using Mesh::FaceIsTrueInterior` just changes the access protection for the base method, so is equivalent to this forwarding dispatch.

dylan-copeland commented on mesh/submesh/submesh.hpp:
```diff
@@ -140,6 +152,43 @@ class SubMesh : public Mesh
    }
 
+   /**
+    * @brief Get the submesh element corresponding to a parent element. -1 == not present.
```
Should it be `< 0 == not present`? Can there be negative values not equal to `-1`? If not, change `pe < 0` to `pe == -1`?

dylan-copeland commented on mesh/submesh/psubmesh.hpp:
```diff
@@ -342,9 +392,16 @@ class ParSubMesh : public ParMesh
    /// The parent Mesh
    const ParMesh &parent_;
 
+   /// Optional nonconformal submesh. Managed via ncmesh pointer in base class.
```
`ncmesh` -> `pncmesh`?

hughcars replied:
They both point to the same so shouldn't make a difference, but will change.

dylan-copeland commented on mesh/submesh/submesh_utils.hpp:
```diff
@@ -40,15 +43,6 @@ struct UniqueIndexGenerator
- * @param el The element
- * @param attributes The attributes
- */
-bool ElementHasAttribute(const Element &el, const Array<int> &attributes);
```
Can this be kept for backwards compatibility? Even if it is not used anywhere in MFEM, is it possible that an MFEM application uses it?

hughcars replied:
I would posit that this isn't really supposed to be public api, but can easily add:
```
bool ElementHasAttribute(const Element &el, const Array<int> &attributes)
{
   return HasAttribute(el,attributes);
}
```

dylan-copeland replied:
We usually avoid changes to the public API, even when it doesn't make sense, to avoid breaking applications that depend on MFEM. It looks your suggested change would maintain the API.

dylan-copeland commented on mesh/submesh/submesh_utils.hpp:
```diff
@@ -123,6 +117,138 @@ RT GetRootParent(const T &m)
 }
 
+/**
+ * @brief Add boundary elements to the Mesh.
```
`Mesh` -> `SubMesh`

dylan-copeland commented on mesh/submesh/submesh_utils.hpp:
```diff
@@ -123,6 +117,138 @@ RT GetRootParent(const T &m)
+ * @details An attempt to call this function for anything other than SubMesh or ParSubMesh
+ * will result in a linker error as the template is only explicitly instantiated for those
+ * types.
+ * @param mesh The Mesh to add boundary elements to.
```
`Mesh` -> `SubMesh`

dylan-copeland commented on mesh/submesh/submesh_utils.hpp:
```diff
@@ -123,6 +117,138 @@ RT GetRootParent(const T &m)
+ * @param lface_to_boundary_attribute Map from local faces in the submesh to boundary
+ * attributes. Only necessary for interior boundary attributes of volume submeshes, where
+ * the face owning the attribute might be on a neighboring rank.
+ * @tparam MeshT The SubMesh type, options SubMesh and ParSubMesh.
```
`MeshT` -> `SubMeshT`

dylan-copeland commented on mesh/submesh/ncsubmesh.hpp:
```diff
@@ -0,0 +1,118 @@
+public:
+   using From = SubMesh::From;
+   /**
+    * @brief Get the parent Mesh object
```
Just use `///`?

dylan-copeland commented on mesh/submesh/ncsubmesh.hpp:
```diff
@@ -0,0 +1,118 @@
+   }
+
+   /**
+   * @brief Check if Mesh @a m is a SubMesh.
```
`Mesh` -> `NCMesh`

dylan-copeland commented on mesh/submesh/ncsubmesh.hpp:
```diff
@@ -0,0 +1,118 @@
+   NCSubMesh(SubMesh& submesh, const NCMesh &parent, From from,
+             const Array<int> &attributes);
+
+   /// The parent Mesh. Not owned.
```
`Mesh` -> `NCMesh`

dylan-copeland commented on tests/unit/mesh/mesh_test_utils.hpp:
```diff
@@ -98,9 +120,33 @@ Mesh OrientedTriFaceMesh(int orientation, bool add_extbdr = false);
 
 
+/**
+ * @brief Helper to refine a single element attached to a boundary attributes
```
attributes -> attribute

hughcars commented on general/array.hpp:
```diff
@@ -204,6 +212,8 @@ class Array
    /// Delete the whole array.
    inline void DeleteAll();
 
+   /// Reduces the capacity of the array to exactly match the current size.
```
If a vector is Reserved with an upper bound on memory, then elements added with Append, once you're done if the array will never be extended again in future, you can shrink it down to remove the unneeded memory at the end. That way a) during Append there are no copies to a new array, so Append is fast, b) at all times all entries in the Array are valid and c) Once the final size of the array is known, you can release the unneeded memory. It's equivalent to https://en.cppreference.com/w/cpp/container/vector/shrink_to_fit

dylan-copeland replied:
I agree. My point is that (c) is unnecessary and more costly than keeping the unneeded memory. Of course, that depends on the amount of excess memory being used, and whether it is more important to reduce memory usage or avoid using time to reallocate memory. Hopefully, the reserved memory is not too excessive. If very excessive memory is reserved and then reallocated, that would be inefficient. In cases where that is unavoidable, I suppose it is good to have this new function. The user should carefully decide whether this function is beneficial in a particular situation.

dylan-copeland commented on mesh/submesh/psubmesh.hpp:
```diff
@@ -139,13 +153,51 @@ class ParSubMesh : public ParMesh
+   int GetSubMeshElementFromParent(int pe) const
    {
-      return parent_to_submesh_face_ids_;
+      return (pe < 0 || pe >= parent_to_submesh_element_ids_.Size())
```
Would `pe == -1` be more precise than `pe < 0`, as in the serial class?

hughcars commented on mesh/submesh/submesh_utils.hpp:
```diff
@@ -123,6 +117,144 @@ RT GetRootParent(const T &m)
+   }
+   return false;
+}
+inline bool ElementHasAttribute(const Element &el, const Array<int> &attributes)
```
@dylan-copeland This is the forwarding method I mentioned

hughcars commented:
An example diff for testing this using ex1p
```diff
diff --git a/examples/ex1p.cpp b/examples/ex1p.cpp
index c3240d621..27f9630ac 100644
--- a/examples/ex1p.cpp
+++ b/examples/ex1p.cpp
@@ -129,33 +129,47 @@ int main(int argc, char *argv[])
    //    can handle triangular, quadrilateral, tetrahedral, hexahedral, surface
    //    and volume meshes with the same code.
    Mesh mesh(mesh_file, 1, 1);
+   mesh.EnsureNCMesh(true);
    int dim = mesh.Dimension();
 
    // 5. Refine the serial mesh on all processors to increase the resolution. In
    //    this example we do 'ref_levels' of uniform refinement. We choose
    //    'ref_levels' to be the largest number that gives a final mesh with no
    //    more than 10,000 elements.
-   {
-      int ref_levels =
-         (int)floor(log(10000./mesh.GetNE())/log(2.)/dim);
-      for (int l = 0; l < ref_levels; l++)
-      {
-         mesh.UniformRefinement();
-      }
-   }
 
    // 6. Define a parallel mesh by a partitioning of the serial mesh. Refine
    //    this mesh further in parallel to increase the resolution. Once the
    //    parallel mesh is defined, the serial mesh can be deleted.
-   ParMesh pmesh(MPI_COMM_WORLD, mesh);
-   mesh.Clear();
-   {
-      int par_ref_levels = 2;
-      for (int l = 0; l < par_ref_levels; l++)
-      {
-         pmesh.UniformRefinement();
-      }
-   }
+   auto [pmesh, parent_mesh] = [&](){
+      ParMesh pmesh(MPI_COMM_WORLD, mesh);
+      mesh.Clear();
+
+      pmesh.RandomRefinement(0.5);
+      pmesh.RandomRefinement(0.5);
+      pmesh.RandomRefinement(0.5);
+      pmesh.RandomRefinement(0.5);
+      pmesh.RandomRefinement(0.5);
+      pmesh.RandomRefinement(0.5);
+      pmesh.RandomRefinement(0.5);
+      pmesh.RandomRefinement(0.5);
+
+      return std::make_pair(ParSubMesh::CreateFromBoundary(pmesh, {1,2,3}), pmesh);
+   }();
 
    // 7. Define a parallel finite element space on the parallel mesh. Here we
    //    use continuous Lagrange finite elements of the specified order. If
@@ -275,6 +289,11 @@ int main(int argc, char *argv[])
    //     local finite element solution on each processor.
    a.RecoverFEMSolution(X, b, x);
 
+   H1_FECollection par_fec(order, parent_mesh.Dimension());
+   ParFiniteElementSpace par_fespace(&parent_mesh, &par_fec);
+   ParGridFunction par_sol(&par_fespace);
+   pmesh.Transfer(x, par_sol);
+
    // 15. Save the refined mesh and the solution in parallel. This output can
    //     be viewed later using GLVis: "glvis -np <np> -m mesh -g sol".
    {
@@ -300,6 +319,11 @@ int main(int argc, char *argv[])
       sol_sock << "parallel " << num_procs << " " << myid << "\n";
       sol_sock.precision(8);
       sol_sock << "solution\n" << pmesh << x << flush;
+
+      socketstream sol_sock2(vishost, visport);
+      sol_sock2 << "parallel " << num_procs << " " << myid << "\n";
+      sol_sock2.precision(8);
+      sol_sock2 << "solution\n" << parent_mesh << par_sol << flush;
    }
 
    // 17. Free the used memory.

```
Used to generate some random refinements on tets and hexes
<img width="1254" alt="Screenshot 2024-10-21 at 4 18 43 PM" src="https://github.com/user-attachments/assets/25effee7-cac3-4019-be97-bc09ca1a23ed">
<img width="1217" alt="Screenshot 2024-10-21 at 4 13 42 PM" src="https://github.com/user-attachments/assets/b53f1138-ac72-41f0-94f8-c412c081d66f">
<img width="1217" alt="Screenshot 2024-10-21 at 4 13 28 PM" src="https://github.com/user-attachments/assets/12d417b6-ad52-4166-92a0-e439d2b58e9f">

dylan-copeland commented on mesh/ncmesh.hpp:
```diff
@@ -981,18 +1071,17 @@ class NCMesh
+       The following transformation gives the relation between the reference
+       quad face coordinates (xi, eta) in [0,1]^2, and the fine quad face
+       coordinates (x, y): x = a0*(1-xi)*(1-eta) + a1*xi*(1-eta) + a2*xi*eta +
+       a3*(1-xi)*eta y = b0*(1-xi)*(1-eta) + b1*xi*(1-eta) + b2*xi*eta +
```
The alignment of the 2 equations is lost, and they should be on separate lines.

dylan-copeland commented on mesh/submesh/pncsubmesh.cpp:
```diff
@@ -0,0 +1,157 @@
+
+
+ParNCSubMesh::ParNCSubMesh(ParSubMesh& submesh,
+                           const ParNCMesh &parent, From from, const Array<int> &attributes)
```
A rare long line. Thanks for correcting many other long lines in the `master` branch.

dylan-copeland commented on mesh/submesh/submesh_utils.cpp:
```diff
@@ -227,5 +253,607 @@ Array<int> BuildFaceMap(const Mesh& pm, const Mesh& sm,
+         - Check not top level face
+         - Check for parent of the newly entered element
+            - if not present, add in
+            - if present but different order and this path is non-ambiguous, reorder so consistent with child
```
Long line.

hughcars replied:
Fixed in https://github.com/mfem/mfem/pull/4476/commits/14e511403985758e7c9c8c131907bcad1a3a995b

dylan-copeland commented on mesh/submesh/psubmesh.hpp:
```diff
@@ -64,7 +68,7 @@ class ParSubMesh : public ParMesh
+                                      const Array<int> &domain_attributes);
 
    /**
    * @brief Create a surface ParSubMesh from it's parent.
```
`it's` -> `its` (preexisting typo) in 2 places.

hughcars replied:
Fixed in https://github.com/mfem/mfem/pull/4476/commits/14e511403985758e7c9c8c131907bcad1a3a995b

hughcars commented:
Github isn't allowing me to comment inline:
"Thanks for restoring this function as deprecated. It is moved from protected to public, which I think is harmless but maybe unintended. There are also some new versions of FindFaceNodes. Is there a reason for them to be public, or can all of these functions be protected?"

So I changed it back before the CI reminded me why I changed it. 

Basically the friend statements in `NCSubMesh`
```
   friend void SubMeshUtils::ConstructFaceTree<NCSubMesh>(NCSubMesh &submesh,
                                                          const Array<int> &attributes);
   friend void SubMeshUtils::ConstructVolumeTree<NCSubMesh>(NCSubMesh &submesh,
                                                            const Array<int> &attributes);
```
are insufficient to be allowed to view protected methods on `.parent` member variable. Oddly this compiles just fine on my laptop clang (erroneously), but the CI corrected me: https://github.com/mfem/mfem/actions/runs/11520323875/job/32071367828?pr=4476#step:9:263 

The fix is to roll that method back to public, which I don't mind at all, given the methods are `const` and do not modify any internals, just viewing them.

hughcars commented:
Thank you for the review @dylan-copeland!

jandrej commented on mesh/submesh/submesh_utils.cpp:
```diff
@@ -227,5 +253,607 @@ Array<int> BuildFaceMap(const Mesh& pm, const Mesh& sm,
+} // namespace
+
+template<typename NCSubMeshT>
+void ConstructFaceTree(NCSubMeshT &submesh, const Array<int> &attributes)
```
noice

jandrej commented on mesh/ncmesh.cpp:
I can only trust the tests on this

hughcars commented on mesh/submesh/ncsubmesh.hpp:
```diff
@@ -0,0 +1,118 @@
+
+   /// Mapping from submesh element nc ids (index of the array), to the parent
+   /// element ids. If from a boundary, these map to faces in the parent.
+   Array<int> parent_element_ids_;
```
The `NCMesh::Elements` are different to those of the mesh, so the map here has to be different to the one in the `SubMesh`, and there has to be duplication across `NCSubMesh` and `ParNCSubMesh` because of the weird inheritance/composition structure of Mesh <-> NCMesh. 

I removed the extra unneeded member variables in https://github.com/mfem/mfem/pull/4476/commits/9f708051daefaacde29fc2ba0447050eeb625fa2

hughcars commented on mesh/mesh.hpp:
```diff
@@ -1102,6 +1112,15 @@ class Mesh
+   /**
+    * @brief Clear the boundary element to edge map.
+    */
+   void RemoveBoundaryElementToEdge()
```
Changed in https://github.com/mfem/mfem/pull/4476/commits/9f708051daefaacde29fc2ba0447050eeb625fa2

jandrej commented:
Thank you for the contribution!

tzanio commented:
Merged in `next` for testing...

tzanio commented:
Please describe in `CHANGELOG`

tzanio commented:
Re-merged in `next` for testing...

v-dobrev commented:
In the nightly tests, there was a failure with gcc 8.3.1, so I pushed a small tweak to work around that: 564e2ff58e.

tzanio commented:
Re-merged in `next` for testing...

tzanio reviewed (approved):
Thank you so much for this important contribution, @hughcars! 🚀
